### PR TITLE
Asynchronously deploy from slack

### PR DESCRIPTION
**Pull request by therealplato**
This PR adds `deploy` to the commands handled by Slack. Deployments are asynchronous: If the user provided a syntactically valid command, Slackbot whispers "Deployment started" to the user. Later, the deployment result gets posted publicly to the channel.

The `/deploy/:PlaybookID/:InstanceID` route is not changed from a synchronous deployment.

Before merging we should check out the deployment tests and see if we can mock kubernetes, the `/services` tests take ~35 seconds

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -223,35 +225,39 @@ func (s *Server) postCommand(c *gin.Context) {
...
 	is := services.NewInstanceService(s.store)
 	i, err := is.Show(pID, ID)
 	if err != nil {
 		return nil, err
 	}
 
 	ds := services.NewDeploymentService(s.store, s.playbooks, s.manifests)
-	err = ds.Deploy(i)
+
+	// will block:
```
Comment (poopoothegorilla) [inline server/server.go]: remove comment before commit

File: services/deployment.go [hunk 2]
```diff
@@ -41,29 +41,40 @@ func vars(i *instance.Instance) map[string]string {
 	return vs
 }
 
-// Deploy deploys a playbook
-func (d *DeploymentService) Deploy(i *instance.Instance) error {
+// DeployAndNotify attempts to deploy a playbook. It reports success or failure
```
Comment (poopoothegorilla) [inline services/deployment.go]: Isaac: "It attempts to deploy an instance not playbook"

File: services/deployment.go [hunk 3]
```diff
@@ -72,43 +83,46 @@ func (d *DeploymentService) Deploy(i *instance.Instance) error {
...
 	}
 
-	err = deployer.Deploy()
-	if err != nil {
-		glog.Errorf("Deploying %s/%s failed: %s\n", i.PlaybookID, i.ID, err.Error())
+	err1 := deployer.Deploy()
+	if err1 != nil {
+		// Mark the instance as problematic:
 		i.Status = instance.StatusError
+		err3 := d.repo.Save(i)
```
Comment (poopoothegorilla) [inline services/deployment.go]: fix numbrs

File: services/slack.go [hunk 4]
```diff
@@ -15,11 +16,40 @@ type SlackCommand interface {
...
+	if err != nil {
+		glog.Error(err)
+	}
+
+	ds := NewDeploymentService(store.New(), deployment.AllPlaybooks, AllManifests)
+
+	i, err := c.is.Show(c.pID, c.ID)
+	if err != nil {
+		glog.Errorf("Tried to deploy missing instance %s/%s", i.PlaybookID, i.ID)
+		return fmt.Sprintf("Failed to deploy instance %s/%s: Instance not found", i.PlaybookID, i.ID), err
```
Comment (poopoothegorilla) [inline services/slack.go]: clean up idea: string combo

File: services/slack.go [hunk 5]
```diff
@@ -15,11 +16,40 @@ type SlackCommand interface {
...
+	}
+
+	glog.Infof("Asynchronously deploying %s/%s...", i.PlaybookID, i.ID)
```
Comment (poopoothegorilla) [inline services/slack.go]: move inside 42 `go func()`

File: services/deployment.go [hunk 6]
```diff
@@ -41,29 +41,40 @@ func vars(i *instance.Instance) map[string]string {
...
 }
 
-// Deploy deploys a playbook
-func (d *DeploymentService) Deploy(i *instance.Instance) error {
+// DeployAndNotify attempts to deploy a playbook. It reports success or failure
+// through the notification service as well as returning an error.
+func (d *DeploymentService) DeployAndNotify(i *instance.Instance) error {
 	playbook, ok := d.playbooks[i.PlaybookID]
 	if !ok {
+		msg := fmt.Sprintf("Can't deploy %s/%s: Playbook missing", i.PlaybookID, i.ID)
```
Comment (poopoothegorilla) [inline services/deployment.go]: DRY

File: services/slack.go [hunk 7]
```diff
@@ -15,11 +16,40 @@ type SlackCommand interface {
...
+
+	go func() {
+		err := ds.DeployAndNotify(i)
+		if err != nil {
+			glog.Errorf("Slack command failed to deploy instance %s/%s", i.PlaybookID, i.ID)
```
Comment (poopoothegorilla) [inline services/slack.go]: log error?

File: services/deployment.go [hunk 8]
```diff
@@ -72,43 +83,46 @@ func (d *DeploymentService) Deploy(i *instance.Instance) error {
...
-		Attachments: []notification.Attachment{
-			{
-				Text: fmt.Sprintf("Instance was deployed: %s %s.", i.PlaybookID, i.ID),
-			},
-		},
+func notify(i *instance.Instance, msg string) {
+	m := notification.NewMessage(false, msg)
+	err := m.Send()
+	if err != nil {
+		glog.Infof("Failed to send notification from DeploymentService:\n%+v\n", m)
```
Comment (macat) [inline services/deployment.go]: this should be a `glog.Warningf`

**Comment by macat** [comment]
Other than that log line, this LGTM

**Pull request merged**